active/0024--message-validation-and-transform.md

```diff
@@ -0,0 +1,162 @@
+# Message validation and transform
+
+## Changelog
+
+* 2020-10-21: @zmstone Initial draft
+
+## Abstract
+
+A new feature for EMQX v5 to validate, transform, and filter MQTT messages.
+
+## Motivation
+
+### Data validation
+
+EMQX has no native support for data validation.
+
+Technically, by using EMQX rules engine, it is possible to validate the incoming
+MQTT messages before sending the message off to data bridges.
+
+For instance, one can write an rules engine SQL like below to
+drop messages which are not avro encoded and also when the 'val' field is not greater than 10.
+
+```
+SELECT schema_decode(my_avro_schema, payload) as decoded from t/# where decoded.val > 10
+```
+
+However, currently rules are only used in data bridges, but it does not interfere
+MQTT message pub/sub.
+For example, if there is subscriber subscribing to `t/#` it would still receive the avro encoded message.
+
+If we also want to stop the message being published, then there would be a need for a message republish action,
+and the publisher and subscriber will likely be forced to use different topics.
+e.g. publisher publishes to `t1/...`, republish action send it to `t2/...` for subscribers to receive.
+
+This can work, but not easy to use.
+
+### Message transform
+
+Same as data validation, if message is to be transformed using rules engine SQL,
+it would require a republish action.
+
+## Design
+
+Develop a data processing feature which can support:
+
+- Message validation
+- Message transform
+- Message filtering
+
+This data processing step should happen after authorization and before rules engine.
+Probably can make use of the 'message.publish' with a hook priority which
+positions the callback after authorization, but before data bridges.
+
+### High level design
+
+* Introduce 2 different data processing types: `validation` and `tranform`.
+* Make use of the rules engine SQL, but without the `FROM` caluse as it's always the MQTT message being th input.
+
+The each data process rule should have a unique ID. And each rule should have a input topic-filter name which
+can be used to build a topic index for quick look-up.
+
+The common parts can be described as hocon config below:
+
+```
+{
+    id = id1 # unique id assigned by user
+    topic = 't/1' # input
+    sql = "SELECT ....." # processing
+    type = validation | transform
+}
+
+```
+
+### Message Validation
+
+```
+{
+    id = id1
+    topic = 't/1'
+    type = validation
+    action_if_invalid = disconnect # or 'drop'
```

Comment:
If we use `message filtering` as common denominator for both validation and filtering, we could have instead `action_if_passed` and `action_if_failed`.